### PR TITLE
scheduler: let balance region do not always select same source and target

### PR DESCRIPTION
**Pull request by nolouch**
Signed-off-by: nolouch <nolouch@gmail.com>

### What problem does this PR solve? 
the balance region scheduler may not work in some cornor case, such as:
```
// 11 regions distributed across 5 stores.
//| region_id | leader_store | follower_store | follower_store |
//|-----------|--------------|----------------|----------------|
//|     1     |       1      |        2       |       3        |
//|     2     |       1      |        2       |       3        |
//|     3     |       1      |        2       |       3        |
//|     4     |       1      |        2       |       3        |
//|     5     |       1      |        2       |       3        |
//|     6     |       1      |        2       |       3        |
//|     7     |       1      |        2       |       4        |
//|     8     |       1      |        2       |       4        |
//|     9     |       1      |        2       |       4        |
//|    10     |       1      |        4       |       5        |
//|    11     |       1      |        4       |       5        |
// and the space of last store 5 is very small, about 5 * regionsize
// the source region is more likely distributed in store[1, 2, 3].
```
It may always select same source store and target store, but `shouldBalance` cannot pass.

### What is changed and how it works?
record the source and target for a while and filter them in next schedule.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: server/schedulers/balance_region.go [hunk 1]
```diff
@@ -205,3 +213,83 @@ func (s *balanceRegionScheduler) hasPotentialTarget(cluster schedule.Cluster, re
...
+	filter := make(hintsStoreFilter)
+	for _, target := range cluster.GetStores() {
+		if h.filter(source, target) {
+			filter[target.GetID()] = true
+		}
+	}
+	return filter
+}
+
+type hintsStoreFilter map[uint64]bool
```
Comment (disksing) [inline server/schedulers/balance_region.go]: Seems better to put it to `schedule`?

File: server/schedulers/balance_region.go [hunk 2]
```diff
@@ -31,14 +32,19 @@ func init() {
...
+	hintsStoreTTL            = time.Minute
+	hintsStoreCountThreshold = 1000
+)
 
 type balanceRegionScheduler struct {
 	*baseScheduler
 	selector     *schedule.BalanceSelector
 	taintStores  *cache.TTLUint64
 	opController *schedule.OperatorController
+	hintsCounter *hintsStoreBuilder
```
Comment (disksing) [inline server/schedulers/balance_region.go]: Can we remove `taintStores` as `hitsFilter` seems to be a better solution for the same problem?

File: server/schedule/filters.go [hunk 3]
```diff
@@ -412,3 +412,60 @@ func (f StoreStateFilter) filterMoveRegion(opt Options, store *core.StoreInfo) b
 	}
 	return false
 }
+
+// BlacklistType the type of BlackListStore Filter.
+type BlacklistType int
+
+// some flags about blacklist type.
+const (
+	SignSource BlacklistType = iota + 1
```
Comment (rleungx) [inline server/schedule/filters.go]: We need a better name or some comments for it.

File: server/schedule/filters.go [hunk 4]
```diff
@@ -412,3 +412,60 @@ func (f StoreStateFilter) filterMoveRegion(opt Options, store *core.StoreInfo) b
...
+	SignTarget
+	SignSourceAndTarget
+)
+
+// BlacklistStoreFilter filter the store according to the blacklist.
```
Comment (rleungx) [inline server/schedule/filters.go]: filters

File: server/schedule/filters.go [hunk 5]
```diff
@@ -412,3 +412,60 @@ func (f StoreStateFilter) filterMoveRegion(opt Options, store *core.StoreInfo) b
...
+type BlacklistStoreFilter struct {
+	blacklist map[uint64]struct{}
+	flag      BlacklistType
+}
+
+// NewBlacklistStoreFilter create a blacklist filter.
```
Comment (rleungx) [inline server/schedule/filters.go]: creates

File: server/schedulers/balance_region.go [hunk 6]
```diff
@@ -28,30 +30,35 @@ func init() {
 	})
 }
 
-// balanceRegionRetryLimit is the limit to retry schedule for selected store.
-const balanceRegionRetryLimit = 10
+const (
+	// balanceRegionRetryLimit is the limit to retry schedule for selected store.
+	balanceRegionRetryLimit = 10
+	hitsStoreTTL            = 5 * time.Minute
+	// ScheduleIntervalFactor default is 1.3 , and MinScheduleInterval is 10ms,
```
Comment (rleungx) [inline server/schedulers/balance_region.go]: I think we need more clear notations here.

File: server/schedulers/balance_region.go [hunk 7]
```diff
@@ -174,31 +171,92 @@ func (s *balanceRegionScheduler) transferPeer(cluster schedule.Cluster, region *
...
-		if !store.IsUp() || store.DownTime() > cluster.GetMaxStoreDownTime() {
-			continue
+func (h *hitsStoreBuilder) getKey(source, target *core.StoreInfo) string {
+	var key string
+	if source == nil || source == nil && target == nil {
+		return ""
+	} else if target == nil {
+		key = fmt.Sprintf("s%d", source.GetID())
+	} else {
+		key = fmt.Sprintf("t%d<-s%d", source.GetID(), target.GetID())
```
Comment (rleungx) [inline server/schedulers/balance_region.go]: Do you mean `fmt.Sprintf("t%d<-s%d", target.GetID(), source.GetID())`?

File: server/schedulers/balance_region.go [hunk 8]
```diff
@@ -174,31 +171,92 @@ func (s *balanceRegionScheduler) transferPeer(cluster schedule.Cluster, region *
...
+
+func (h *hitsStoreBuilder) filter(source, target *core.StoreInfo) bool {
+	key := h.getKey(source, target)
+	if item, ok := h.hits[key]; ok && key != "" {
+		if time.Since(item.lastTime) > h.ttl {
+			delete(h.hits, key)
 		}
-		if !shouldBalance(cluster, source, store, region, core.RegionKind, opInfluence) {
-			continue
+		if time.Since(item.lastTime) < h.ttl && item.count >= h.threshold {
```
Comment (rleungx) [inline server/schedulers/balance_region.go]: What if `time.Since(item.lastTime) == h.ttl`?

File: server/schedule/filters.go [hunk 9]
```diff
@@ -412,3 +412,60 @@ func (f StoreStateFilter) filterMoveRegion(opt Options, store *core.StoreInfo) b
...
 }
+
+// BlacklistType the type of BlackListStore Filter.
+type BlacklistType int
+
+// some flags about blacklist type.
+const (
+	SignSource BlacklistType = iota + 1
+	SignTarget
+	SignSourceAndTarget
```
Comment (rleungx) [inline server/schedule/filters.go]: When do we use it?

Comment (nolouch) [reply server/schedule/filters.go]: We do not used it now.

File: server/schedule/filters.go (hunk 3 above)
Comment (disksing) [reply server/schedule/filters.go]: What does `Sign` mean?

File: server/schedule/filters.go (hunk 9 above)
Comment (disksing) [reply server/schedule/filters.go]: how about using bit flags?
```
source = 1 << iota
target
```
Then `if flag&source` or `if flag&target`

File: server/schedule/filters.go [hunk 10]
```diff
@@ -412,3 +412,61 @@ func (f StoreStateFilter) filterMoveRegion(opt Options, store *core.StoreInfo) b
...
 	return false
 }
+
+// BlacklistType the type of BlackListStore Filter.
+type BlacklistType int
+
+// some flags about blacklist type.
+const (
+	// blacklist associated with the source.
+	SignSource BlacklistType = 1 << iota
```
Comment (disksing) [inline server/schedule/filters.go]: em, what does `sign` mean here?

File: server/schedule/filters.go [hunk 11]
```diff
@@ -412,3 +412,61 @@ func (f StoreStateFilter) filterMoveRegion(opt Options, store *core.StoreInfo) b
...
+		return false
+	}
+	return f.filter(store)
+}
+
+func (f *BlacklistStoreFilter) filter(store *core.StoreInfo) bool {
+	if _, ok := f.blacklist[store.GetID()]; ok {
+		return true
+	}
+	return false
```
Comment (disksing) [inline server/schedule/filters.go]:
```
_, ok := f.blacklist[store.GetID()]
return ok
```

File: server/schedulers/balance_region.go [hunk 12]
```diff
@@ -179,31 +179,92 @@ func (s *balanceRegionScheduler) transferPeer(cluster schedule.Cluster, region *
...
 
-	for _, store := range cluster.GetStores() {
-		if schedule.FilterTarget(cluster, store, filters) {
-			continue
-		}
-		if !store.IsUp() || store.DownTime() > cluster.GetMaxStoreDownTime() {
-			continue
+func (h *hitsStoreBuilder) getKey(source, target *core.StoreInfo) string {
+	var key string
+	if source == nil || source == nil && target == nil {
```
Comment (disksing) [inline server/schedulers/balance_region.go]: ??

File: server/schedulers/balance_region.go [hunk 13]
```diff
@@ -179,31 +179,92 @@ func (s *balanceRegionScheduler) transferPeer(cluster schedule.Cluster, region *
...
 }
+
+func (h *hitsStoreBuilder) miss(source, target *core.StoreInfo) {
+	key := h.getKey(source, target)
+	if _, ok := h.hits[key]; ok && key != "" {
+		delete(h.hits, key)
+	}
+}
+
+func (h *hitsStoreBuilder) hit(source, target *core.StoreInfo) {
```
Comment (shafreeck) [inline server/schedulers/balance_region.go]: I think the name `hit` and `miss` are really confused here. It is different from the meaning `hit en entry` or `miss an entry` of the cache.
BTW, why not use the `server/cache` directly?

Comment (nolouch) [reply server/schedulers/balance_region.go]: Do you have any better suggestion name? `server/cache` not suitable for this situation, we need to build a `Filter` with the `ttl` and hit `count`.

**Comment by nolouch** [comment]
PTAL @disksing @shafreeck

File: server/schedulers/balance_test.go [hunk 14]
```diff
@@ -461,6 +460,65 @@ func (s *testBalanceRegionSchedulerSuite) TestReplicas5(c *C) {
...
+	tc.AddRegionStore(4, 5)
+	tc.AddRegionStore(5, 2)
+	tc.AddLeaderRegion(1, 1, 2, 3)
+	tc.AddLeaderRegion(2, 1, 2, 3)
+
+	c.Assert(sb.Schedule(tc)[0], NotNil)
+	// if the space of store 5 is normal, we can balance region to store 5
+	testutil.CheckTransferPeer(c, sb.Schedule(tc)[0], schedule.OpBalance, 1, 5)
+
+	// the used size of  store 5 reach (highSpace, lowSpace)
```
Comment (rleungx) [inline server/schedulers/balance_test.go]:
```suggestion
	// the used size of store 5 reach (highSpace, lowSpace)
```

File: server/schedulers/balance_test.go [hunk 15]
```diff
@@ -461,6 +460,65 @@ func (s *testBalanceRegionSchedulerSuite) TestReplicas5(c *C) {
...
+	tc.PutStore(store5)
+
+	// the scheduler always pick store 1 as source store,
+	// and store 5 as target store, but cannot pass `shouldBalance`.
+	c.Assert(sb.Schedule(tc), IsNil)
+	// hits the store many times
+	for i := 0; i < 1000; i++ {
+		sb.Schedule(tc)
+	}
+	// now filte the store 5, and can transfer store 1 to store 4
```
Comment (rleungx) [inline server/schedulers/balance_test.go]: filter

File: server/schedulers/balance_test.go [hunk 16]
```diff
@@ -375,22 +371,22 @@ func (s *testBalanceRegionSchedulerSuite) TestReplicas3(c *C) {
...
+		sb.Schedule(tc)
+	}
+	hit := sb.(*balanceRegionScheduler).hitsCounter
+	c.Assert(hit.buildSourceFilter(tc).FilterSource(tc, tc.GetStore(1)), IsTrue)
+	c.Assert(hit.buildSourceFilter(tc).FilterSource(tc, tc.GetStore(2)), IsFalse)
+	c.Assert(hit.buildSourceFilter(tc).FilterSource(tc, tc.GetStore(3)), IsFalse)
 
 	// Store 4 has smaller region score than store 2.
 	tc.AddLabelsStore(4, 2, map[string]string{"zone": "z1", "rack": "r2", "host": "h1"})
+	fmt.Println(sb.Schedule(tc))
```
Comment (rleungx) [inline server/schedulers/balance_test.go]: remove this line?

**Comment by nolouch** [comment]
Ping

**Comment by disksing** [comment]
PTAL @shafreeck @rleungx

File: server/schedulers/balance_region.go (hunk 13 above)
Comment (shafreeck) [reply server/schedulers/balance_region.go]: Maybe just call them put/remove? @nolouch 

@disksing  How do you think?

Comment (disksing) [reply server/schedulers/balance_region.go]: Seems `put/remove` is better.

**Pull request merged**